### ISSUE TITLE
More plug&play

### ISSUE THREAD

**Issue by MagicRB** [kabergstrom/prefab#1]
Could this perhaps more plug&play? Right now its a lot of code to get it working. Most of which makes absolutely no sense to me

**Comment by kabergstrom** [comment]
Yes! Absolutely, right now it's just a proving ground and I didn't intend for it to be used by anyone else, but if you have similar goals I'd love to collaborate of course.
You can maybe read some more about what I'm trying to do here: https://github.com/amethyst/amethyst/issues/1894

Atelier-assets is a game engine agnostic asset system that I will try to integrate this prefab system into. 

It would help to hear some about how you want to use a prefab system? What's your ideal library?

**Comment by MagicRB** [comment]
I do think I have similar goals, I want the engine I'm making to be fully data-driven, I not sure if my terminology is correct, basically you would make a game in the engine by creating prefabs and lua code. The ideal API I'm looking for is something resembling
```rust
#[derive(Copy, Clone, ..., PartialEq, Serialize, Deserialize, SomePrefabMarker)]
struct Transform {
    ...
}

#[derive(Copy, Clone, ..., PartialEq, Serialize, Deserialize)]
struct Runtime {
    ...
}

let serializable = world.serialize_entity::<SomePrefabMarker>(entity);
let json_representation = serde_json::to_string(serializable); // Serializes only Transform, becuase Runtime does not derive SomePrefabMarker
let serializable = world.serialize_entity::<()>(entity)
let json_representation_2 = serde_json::to_string(serializable); // Serializes Transform and Runtime
let saved_world = serde_json::to_string(world.serialize_all::<()>()); // Serializes everything since no marker was provided
```

EDIT: I'm sure I'm missing `.unwrap()`s and maybe some other things, just ignore the mistakes

**Comment by MagicRB** [comment]
On the note of collaboration. I'd love to help, though I'm not sure I'll be able to, but I'll try my best

**Comment by kabergstrom** [comment]
Glad to hear! It seems we are pretty aligned.

As for this specific case, I think we can separate it into two problems:
1. Which components/tags can we serialize/deserialize? 

We need to generate the appropriate code for this and register the components/tags somewhere. I intend to implement two derive macros similar to what you proposed, probably calling them `PrefabComponent` and `PrefabTag`. They would use the `inventory` crate to register themselves globally within the program.

Even currently, unregistered components/tags are not serialized when serializing the World, even if they implement Serialize/Deserialize. This is because ComponentTypeId/TagTypeId use `std::any::TypeId`, which is not a stable identifier: it may change every compile. I generally use the `type-uuid` crate as a persistent identifier for types that remains stable across compiles. Hence, the derive macro needs to map a persistent type identifier to a ComponentTypeId/TagTypeId and its corresponding Serialize/Deserialize implementations.

2. Which entities should be serialized when serializing a World?

I think adding an optional `Query` to the serde implementation for `legion` can resolve your concerns - would you agree?
Basically, you would create a query and pass it to the `serialize_with_query` function to filter which entities get serialized.

**Comment by MagicRB** [comment]
Yeah, in my previous post I got entity filtering and component filtering mixed up, sorry for that.

1. Whenever a Component or Tag is to be serialized it should be checked if it implements `PrefabComponent` or `PrefabTag` respectively.

2. Entity filtering should be applied when serializing a whole `World`, where as you suggested, it can be implemented using `Query`. Component level filtering should also be applied using `inventory`, `PrefabComponent`, PrefabTab`.

Though I'm not sure how well will this work with runtime loadable "modules"/"plug-ins"/"mods". If the `inventory` crate can handle it, it should work.

TL;DR I agree with everything you suggested. Only the paragraph just above this one is adding something

**Comment by kabergstrom** [comment]
Great! Sounds workable then. Do you plan on integrating atelier-assets in your game engine? My next point of action after the derive macro is to implement the full prefab system that leverage atelier-assets to load referenced prefabs, field overrides and entity data, then use atelier-assets' build pipeline to "bake" the prefab reference graph to a simple `legion::World` that can be loaded by the game engine runtime.

Regarding point 1, it's generally impossible to detect whether a type implements a trait or not, as Rust does not implement any form of type reflection. Explicit registration of a type is thus required. For the case of dynamically loaded modules, the type registry would probably need to be updated as required by the application. I'll have to think a bit more about it, but it should be generally possible.

**Comment by MagicRB** [comment]
By checking if it implements I meant registration, I'm still stuck in the reflective languages.

I'm not sure if I'll use atelier-assets in my game engine since I'd like the game engine serve as the asset editor, because you want to have two representations of prefabs, that won't be possible, I'm not saying I won't optimize the assets, but i want both states to be understandable and loadable by the game engine. (It might seem like I have no idea how all this will work and it is true, I have none)

**Comment by kabergstrom** [comment]
Are you planning to let users edit assets at runtime? How will you handle storage of changed assets in that case?

**Comment by MagicRB** [comment]
What exactly do you mean? I don't think I understand what you mean by `storage of changed assets`

**Comment by kabergstrom** [comment]
When a user modifies an asset, how/where do you intend to store the data for that asset?

**Comment by MagicRB** [comment]
By asset you mean a texture and such? I'd have everything on a path based system. And I would most likely watch for filesystem changes (i mean operating system file system, the real thing). I know that can be done on Linux, not sure about Windows. I wanted to store all textures in a atlas, (as for sounds and such I have no clue since I have never worked with audio), so when a file gets changed, the engine would detect it, delete the last version from the atlas, place the new one into a free area or do some restructuring. Then it would go through all `Spite` components and mark them for re-evaluation, then another system would actually do the re-evaluation (Most likely just by mutably borrowing, which would trigger a change inside legion, which could then be queried for).  When something triggers a change in the `Sprite` component from within the engine (a script or something) that would automatically mean re-evaluation of the `Sprite`.

I hope what I just wrote makes sense.

**Comment by kabergstrom** [comment]
Sounds like atelier-assets would fit that pretty well. You can run the AssetDaemon on a client computer, it will do filesystem watching for you (on all PC platforms, windows/mac/linux) and invalidate loaded/built assets through a dependency graph.
Atelier-assets doesn't prevent you from having the intermediate and runtime formats be the same type, so you don't *need* to "bake" the prefab before loading. :) I'll look into how to do prefab editing too, as that's going to be very necessary to make prefabs usable from a user point of view.

**Comment by kabergstrom** [comment]
This project is a bit more plug and play now @MagicRB  https://github.com/kabergstrom/prefab/blob/master/legion-prefab/examples/simple.rs

**Comment by MagicRB** [comment]
Great to hear! Once I get more of my engine working, I'll try it and report back, oh BTW I'm going to give atelier assets a shot, though it seems a bit complicated for my usecase
